File: libraries/fary/ary1_cpdlt.f
      SUBROUTINE ARY1_CPDLT( IDCB1, IDCB2, STATUS )
*+
*  Name:
*     ARY1_CPDLT

*  Purpose:
*     Copy delta compression information from one DCB entry to another.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_CPDLT( IDCB1, IDCB2, STATUS )

*  Description:
*     The routine copies the supplemental information describing the
*     compression of a supplied delta array (identified by its DCB entry)
*     to another existing DCB entry. Note, it only copies components of
*     the DELTA array structure that are not also conmponents of a SIMPLE
*     array (e.g. DATA, ORIGIN, etc are not copied).

*  Arguments:
*     IDCB1 = INTEGER (Given)
*        Index to the DCB entry of the array to be copied.
*     IDCB2 = INTEGER (Given)
*        Index to the DCB entry to recieve the copy.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Copyright:
*     Copyright (C) 2010 Science & Technology Facilities Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     DSB: David S Berry (JAC, Hawaii)
*     {enter_new_authors_here}

*  History:
*     1-NOV-2010 (DSB):
*        Original version.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'CNF_PAR'          ! For CNV_PVAL function
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants
      INCLUDE 'ARY_ERR'          ! ARY_ error codes

*  Global Variables:
      INCLUDE 'ARY_DCB'          ! ARY_ Data Control Block
*        DCB_CPX( ARY__MXDCB ) = LOGICAL (Read)
*           Whether the data object is complex.
*        DCB_FRM( ARY__MXDCB ) = CHARACTER * ( ARY__SZFRM ) (Read)
*           Data object storage form.

*  Arguments Given:
      INTEGER IDCB1
      INTEGER IDCB2

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      CHARACTER LOC2*(DAT__SZLOC)! Component locator
      LOGICAL THERE              ! Does the component exist?
*.

*  Check inherited global status.
      IF( STATUS .NE. SAI__OK ) RETURN

*  Ensure that form information is available for the input DCB entry.
      CALL ARY1_DFRM( IDCB1, STATUS )
      IF( STATUS .EQ. SAI__OK ) THEN

*  If the first DCB entry is a delta array then we copy the compression
*  information.
         IF( DCB_FRM( IDCB1 ) .EQ. 'DELTA' ) THEN

*  First do mandatory components. Report an error if any of these does
*  not exist in the input. Otherwise, copy them to the output.

*  ZAXIS
*  -----
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'ZAXIS', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'ZAXIS', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'ZAXIS',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )

            ELSE IF( STATUS .EQ. SAI__OK ) THEN
               STATUS = ARY__DLTIN
               CALL DAT_MSG( 'A', DCB_LOC( IDCB1 )  )
               CALL ERR_REP( ' ', 'The DELTA compressed array '//
     :                       '''^A'' is invalid - the ZAXIS '//
     :                       'component is missing.', STATUS )
            END IF

*  ZDIM
*  ----
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'ZDIM', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'ZDIM', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'ZDIM',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )

            ELSE IF( STATUS .EQ. SAI__OK ) THEN
               STATUS = ARY__DLTIN
               CALL DAT_MSG( 'A', DCB_LOC( IDCB1 ) )
               CALL ERR_REP( ' ', 'The DELTA compressed array '//
     :                       '''^A'' is invalid - the ZDIM '//
     :                       'component is missing.', STATUS )
            END IF

*  VALUE
*  -----
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'VALUE', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'VALUE', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'VALUE',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )

            ELSE IF( STATUS .EQ. SAI__OK ) THEN
               STATUS = ARY__DLTIN
               CALL DAT_MSG( 'A', DCB_LOC( IDCB1 ) )
               CALL ERR_REP( ' ', 'The DELTA compressed array '//
     :                       '''^A'' is invalid - the VALUE '//
     :                       'component is missing.', STATUS )
            END IF

*  FIRST_DATA
*  ----------
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'FIRST_DATA', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'FIRST_DATA', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'FIRST_DATA',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )

            ELSE IF( STATUS .EQ. SAI__OK ) THEN
               STATUS = ARY__DLTIN
               CALL DAT_MSG( 'A', DCB_LOC( IDCB1 ) )
               CALL ERR_REP( ' ', 'The DELTA compressed array '//
     :                       '''^A'' is invalid - the FIRST_DATA '//
     :                       'component is missing.', STATUS )
            END IF


*  FIRST_VALUE
*  ----------
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'FIRST_VALUE', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'FIRST_VALUE', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'FIRST_VALUE',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )

            ELSE IF( STATUS .EQ. SAI__OK ) THEN
               STATUS = ARY__DLTIN
               CALL DAT_MSG( 'A', DCB_LOC( IDCB1 ) )
               CALL ERR_REP( ' ', 'The DELTA compressed array '//
     :                       '''^A'' is invalid - the FIRST_VALUE '//
     :                       'component is missing.', STATUS )
            END IF

*  ZRATIO
*  ------
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'ZRATIO', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'ZRATIO', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'ZRATIO',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )

            ELSE IF( STATUS .EQ. SAI__OK ) THEN
               STATUS = ARY__DLTIN
               CALL DAT_MSG( 'A', DCB_LOC( IDCB1 ) )
               CALL ERR_REP( ' ', 'The DELTA compressed array '//
     :                       '''^A'' is invalid - the ZRATIO '//
     :                       'component is missing.', STATUS )
            END IF

*  Now do optional components. Copy these if they exist in the input, but
*  do not report an error if they do not exist.

*  SCALE
*  -----
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'SCALE', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'SCALE', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'SCALE',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )
            END IF


*  ZERO
*  ----
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'ZERO', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'ZERO', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'ZERO',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )
            END IF

*  REPEAT
*  ------
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'REPEAT', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'REPEAT', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ), 'REPEAT',
     :                        STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )
            END IF

*  FIRST_REPEAT
*  ------------
            CALL DAT_THERE( DCB_LOC( IDCB1 ), 'FIRST_REPEAT', THERE,
     :                      STATUS )
            IF( THERE ) THEN
               CALL DAT_FIND( DCB_LOC( IDCB1 ), 'FIRST_REPEAT', LOC2,
     :                        STATUS )
               CALL DAT_COPY( LOC2,  DCB_LOC( IDCB2 ),
     :                        'FIRST_REPEAT', STATUS )
               CALL DAT_ANNUL( LOC2, STATUS )
            END IF

*  Ensure the storage form is now delta in both the data object and the
*  DCB.
            CALL CMP_MODC( DCB_LOC( IDCB2 ), 'VARIANT', 6, 0, 0,
     :                     STATUS )
            CALL CMP_PUT0C( DCB_LOC( IDCB2 ), 'VARIANT', 'DELTA',
     :                      STATUS )
            DCB_FRM( IDCB2 ) = 'DELTA'

*  Report an error if the input array is not a DELTA array.
         ELSE IF( STATUS .EQ. SAI__OK ) THEN
            STATUS = ARY__FATIN
            CALL ERR_REP( 'ARY1_CPDLT_FORM', 'ARY1_CPDLT: Input '//
     :                    'array is not a DELTA array (internal '//
     :                    'programming error).', STATUS )
         END IF

      END IF

*  Call error tracing routine and exit.
      IF( STATUS .NE. SAI__OK ) CALL ARY1_TRACE( 'ARY1_CPDLT', STATUS )

      END

File: libraries/ary/ary1Cpdlt.c
#include "sae_par.h"
#include "ary1.h"
#include "star/hds.h"
#include "star/cmp.h"
#include "mers.h"
#include "ary_err.h"
#include <string.h>

void ary1Cpdlt( AryDCB *dcb1, AryDCB *dcb2, int *status ) {
/*
*+
*  Name:
*     ary1Cpdlt

*  Purpose:
*     Copy delta compression information from one DCB entry to another.

*  Synopsis:
*     void ary1Cpdlt( AryDCB *dcb1, AryDCB *dcb2, int *status )

*  Description:
*     This function copies the supplemental information describing the
*     compression of a supplied delta array (identified by its DCB entry)
*     to another existing DCB entry. Note, it only copies components of
*     the DELTA array structure that are not also conmponents of a SIMPLE
*     array (e.g. DATA, ORIGIN, etc are not copied).

*  Parameters:
*     dcb1
*        The DCB entry of the array to be copied.
*     dcb2
*        The DCB entry to recieve the copy.
*     status
*        The global status.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Local variables: */
   int there;
   HDSLoc *loc2 = NULL;

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Ensure that form information is available for the input DCB entry. */
   ary1Dfrm( dcb1, status );
   if( *status == SAI__OK ){

/* If the first DCB entry is a delta array then we copy the compression
   information. */
      if( !strcmp( dcb1->form, "DELTA" ) ){

/* First do mandatory components. Report an error if any of these does not
   exist in the input. Otherwise, copy them to the output. */

/* ZAXIS ----- */
         datThere( dcb1->loc, "ZAXIS", &there, status );
         if( there ){
            datFind( dcb1->loc, "ZAXIS", &loc2, status );
            datCopy( loc2, dcb2->loc, "ZAXIS", status );
            datAnnul( &loc2, status );

         } else if( *status == SAI__OK ){
            *status = ARY__DLTIN;
            datMsg( "A", dcb1->loc );
            errRep( " ", "The DELTA compressed array '^A' is invalid - the ZAXIS "
                    "component is missing.", status );
         }

/* ZDIM ---- */
         datThere( dcb1->loc, "ZDIM", &there, status );
         if( there ){
            datFind( dcb1->loc, "ZDIM", &loc2, status );
            datCopy( loc2, dcb2->loc, "ZDIM", status );
            datAnnul( &loc2, status );

         } else if( *status == SAI__OK ){
            *status = ARY__DLTIN;
            datMsg( "A", dcb1->loc );
            errRep( " ", "The DELTA compressed array '^A' is invalid - the ZDIM "
                    "component is missing.", status );
         }

/* VALUE ----- */
         datThere( dcb1->loc, "VALUE", &there, status );
         if( there ){
            datFind( dcb1->loc, "VALUE", &loc2, status );
            datCopy( loc2, dcb2->loc, "VALUE", status );
            datAnnul( &loc2, status );

         } else if( *status == SAI__OK ){
            *status = ARY__DLTIN;
            datMsg( "A", dcb1->loc );
            errRep( " ", "The DELTA compressed array '^A' is invalid - the VALUE "
                    "component is missing.", status );
         }

/* FIRST_DATA ---------- */
         datThere( dcb1->loc, "FIRST_DATA", &there, status );
         if( there ){
            datFind( dcb1->loc, "FIRST_DATA", &loc2, status );
            datCopy( loc2, dcb2->loc, "FIRST_DATA", status );
            datAnnul( &loc2, status );

         } else if( *status == SAI__OK ){
            *status = ARY__DLTIN;
            datMsg( "A", dcb1->loc );
            errRep( " ", "The DELTA compressed array '^A' is invalid - the "
                    "FIRST_DATA component is missing.", status );
         }


/* FIRST_VALUE ---------- */
         datThere( dcb1->loc, "FIRST_VALUE", &there, status );
         if( there ){
            datFind( dcb1->loc, "FIRST_VALUE", &loc2, status );
            datCopy( loc2, dcb2->loc, "FIRST_VALUE", status );
            datAnnul( &loc2, status );

         } else if( *status == SAI__OK ){
            *status = ARY__DLTIN;
            datMsg( "A", dcb1->loc );
            errRep( " ", "The DELTA compressed array '^A' is invalid - the "
                    "FIRST_VALUE component is missing.", status );
         }

/* ZRATIO ------ */
         datThere( dcb1->loc, "ZRATIO", &there, status );
         if( there ){
            datFind( dcb1->loc, "ZRATIO", &loc2, status );
            datCopy( loc2, dcb2->loc, "ZRATIO", status );
            datAnnul( &loc2, status );

         } else if( *status == SAI__OK ){
            *status = ARY__DLTIN;
            datMsg( "A", dcb1->loc );
            errRep( " ", "The DELTA compressed array '^A' is invalid - the ZRATIO "
                    "component is missing.", status );
         }

/* Now do optional components. Copy these if they exist in the input, but
   do not report an error if they do not exist. */

/* SCALE ----- */
         datThere( dcb1->loc, "SCALE", &there, status );
         if( there ){
            datFind( dcb1->loc, "SCALE", &loc2, status );
            datCopy( loc2, dcb2->loc, "SCALE", status );
            datAnnul( &loc2, status );
         }


/* ZERO ---- */
         datThere( dcb1->loc, "ZERO", &there, status );
         if( there ){
            datFind( dcb1->loc, "ZERO", &loc2, status );
            datCopy( loc2, dcb2->loc, "ZERO", status );
            datAnnul( &loc2, status );
         }

/* REPEAT ------ */
         datThere( dcb1->loc, "REPEAT", &there, status );
         if( there ){
            datFind( dcb1->loc, "REPEAT", &loc2, status );
            datCopy( loc2, dcb2->loc, "REPEAT", status );
            datAnnul( &loc2, status );
         }

/* FIRST_REPEAT ------------ */
         datThere( dcb1->loc, "FIRST_REPEAT", &there, status );
         if( there ){
            datFind( dcb1->loc, "FIRST_REPEAT", &loc2, status );
            datCopy( loc2, dcb2->loc, "FIRST_REPEAT", status );
            datAnnul( &loc2, status );
         }

/* Ensure the storage form is now delta in both the data object and the
   DCB. */
         datThere( dcb2->loc, "VARIANT", &there, status );
         if( there ) datErase( dcb2->loc, "VARIANT", status );
         datNew0C( dcb2->loc, "VARIANT", 6, status );
         cmpPut0C( dcb2->loc, "VARIANT", "DELTA", status );
         strcpy( dcb2->form, "DELTA" );

/* Report an error if the input array is not a DELTA array. */
      } else if( *status == SAI__OK ){
         *status = ARY__FATIN;
         errRep( " ", "ary1Cpdlt: Input array is not a DELTA array (internal "
                 "programming error).", status );
      }

   }

/* Call error tracing routine and exit. */
   if( *status != SAI__OK ) ary1Trace( "ary1Cpdlt", status );

}
